src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.cs

```diff
@@ -38,6 +38,15 @@ public static unsafe partial class NativeMemory
             return AllocZeroed(byteCount, elementSize: 1);
         }
 
+        /// <summary>Clears a block of memory.</summary>
+        /// <param name="ptr">A pointer to the block of memory that should be cleared.</param>
+        /// <param name="byteCount">The size, in bytes, of the block to clear.</param>
+        [CLSCompliant(false)]
+        public static unsafe void ZeroMemory(byte* ptr, nuint byteCount)
+        {
+            SpanHelpers.ClearWithoutReferences(ref *ptr, byteCount);
```

Comment:
Do we want to do the check for non-NullRef or a `byteCount` of 0? `ClearWithoutReferences` only checks for non-zero `byteCount` and just uses potentially NullRef.

Reply:
Was just discussing this with @tannergooding on Discord, yeah I'm just not sure on what to do there just yet. We could either say that this mirrors `memset`, which is undefined in this case, and therefore a NRE is fine, or we might want to validate that and throw a better exception. I guess there's an argument for both between usability and perf. Happy to go with whatever you think is more appropriate here 🙂

Reply:
The `NativeMemory.*` APIs are generally meant to be equivalent to "thin" wrappers over the corresponding `C Runtime` APIs.

In this case, it would be a "thin" wrapper over `ZeroMemory` or `memset(ptr, 0, size)`.

The CRT APIs don't have a "standard" behavior here, but do frequently allow `null` alongside `0`, treating it as a `nop`. So it may be beneficial to do the same here.

Reply:
(It may even be beneficial to just guarantee that this calls `memset(ptr, 0, size)` and to leave it at that; it gives a well-defined behavior that can then be used to build up higher level "general-purpose" APIs).

Reply:
> but do frequently allow null alongside 0, treating it as a nop.

This makes sense to me.

Reply:
Just so I understand:

- `SpanHelpers.ClearWithoutReferences` already no-ops when the size is 0, so that part is already handled I believe.
- What should we do here then if the pointer is null and the size is > 0? `ArgumentOutOfRangeException`?
- Doing so would mean we'd always have two branches (`ptr is null && size > 0`), is that a concern for perf?

Reply:
> What should we do here then if the pointer is null and the size is > 0? ArgumentOutOfRangeException?

Sounds good to me.

> Doing so would mean we'd always have two branches (ptr is null && size > 0), is that a concern for perf?

I've no sense of how much this is an issue. I'd like to see how impactful that check is in practice. For example, if size is larger than 128 is the null check even observable? Seems like a BenchmarkDotNet test could be written that determines where the cost becomes noticeable and then we can see if that is below the assumed common threshold.

Reply:
> The existing Buffer.MemoryCopy API that this API is augmenting does not have any explicit argument checks.

Ah ha. That is what I was looking for, an existing API that could help justify not checking was okay. Sounds good. I'm okay with keeping it as is then.

Reply:
Should I just add a unit test to validate that calling `ZeroMemory` with a `null` pointer and a `0` length doesn't throw, and then just leave the case when `null` is passed with a non-zero length undefined? I guess we can add some remarks for extra clarity on the behavior of the API in these two specific scenarios?

Reply:
> Should I just add a unit test to validate that calling ZeroMemory with a null pointer and a 0 length doesn't throw, and then just leave the case when null is passed with a non-zero length undefined? 

Sounds good to me.

Reply:
Done in https://github.com/dotnet/runtime/pull/69500/commits/8e01c6fce3748bfa431bb0b88b3dffa16e346e1c and https://github.com/dotnet/runtime/pull/69500/commits/7ea37a37bd291609b8d48a6a30f0cb8eb8a27132 👍